README.md

```diff
@@ -0,0 +1,178 @@
+# Go Middleware Generator
+
+This tool can generate a generic middleware, that injects hooks around
+the invocations of given interface' methods.
+
+# Description
+
+The generated middleware wraps around another implementation of the same interface
+and calls specific hooks at specific points in the execution.
+
+Given the following interface and `go:generate` command
+```go
+//go:generate go run github.com/Makpoc/gomigen/cmd/gomigen . Interface
+type Interface interface {
+    ProcessOne(string) (string, error)
+    ProcessTwo(context.Context, int) bool
+}
+```
+
+running `go:generate` will produce the following Middleware:
+
+* Struct:
+
+The generated struct implements the Interface
+
+```go
+type InterfaceMiddleware struct {
+	next interfaces.Interface
+	hook types.Hook
+}
+```
+
+* Constructor
+
+The middleware constructor wraps another implementation of that interface
+and also accepts a hook implementation, that will be called when a method of
+this middleware is called.
+
+```go
+func NewInterfaceMiddleware(
+	next interfaces.Interface,
+	hook types.Hook,
+) *InterfaceMiddleware {
+	return &InterfaceMiddleware{
+		next: next,
+		hook: hook,
+	}
+}
+```
+
+* Methods
+
+Each generated method calls the OnEntry hook, then calls the wrapped Interface
+implementation and finally calls the OnExit hook.
+
+Depending on that method arguments and return parameters the hooks and the
+wrapped method receive different values. See the inlined comments for notable
+differences.
+
+```go
+func (mw *InterfaceMiddleware) ProcessOne(arg0 string) (string, error) {
+	methodInfo := types.MethodInfo{
+		Package:   "<full-package-path>",
+		Interface: "Interface",
+		Method:    "ProcessTwo",
+	}
+
+	// ProcessOne doesn't accept context so one is created
+	ctx := context.Background()
+
+	ctx = mw.hook.OnEntry(ctx, methodInfo)
+	res0, res1 := mw.next.ProcessOne(arg0)
+	// ProcessOne returns an error as last return parm and it's passed to the OnExit hook
+	mw.hook.OnExit(ctx, methodInfo, res1)
+	return res0, res1
+}
+
+func (mw *TwoMethodsWithArgAndReturnMiddleware) ProcessTwo(arg0 context.Context, arg1 int) bool {
+	methodInfo := types.MethodInfo{
+		Package:   "<full-package-path>",
+		Interface: "Interface",
+		Method:    "ProcessTwo",
+	}
+
+	// ProcessTwo accepts a context and it is used
+	ctx := arg0
+
+	ctx = mw.hook.OnEntry(ctx, methodInfo)
+	res0 := mw.next.ProcessTwo(ctx, arg1)
+	// ProcessTwo doesn't return an error type so OnExit always receives nil
+	mw.hook.OnExit(ctx, methodInfo, nil)
+	return res0
+}
+
+```
+
+All hooks accept a `Context` they can extract values from as well as `MethodInfo`, containing
+information about the method being invoked. Some hooks accept additional parameters also
```

Comment:
The sentence `Some hooks accept ....` feels incomplete. I personally could not understand what it was trying to convey.

Reply:
I've reworded the readme and dropped that sentence.